xarray/core/datatree.py

```diff
@@ -447,45 +510,28 @@ def ds(self, data: Dataset | DataArray | None = None) -> None:
         # Known mypy issue for setters with different type to property:
         # https://github.com/python/mypy/issues/3004
         ds = _coerce_to_dataset(data)
+        self._replace_node(ds)
 
-        _check_for_name_collisions(self.children, ds.variables)
-
-        self._replace(
-            inplace=True,
-            variables=ds._variables,
-            coord_names=ds._coord_names,
-            dims=ds._dims,
-            indexes=ds._indexes,
-            attrs=ds._attrs,
-            encoding=ds._encoding,
-        )
-        self._close = ds._close
-
-    def _pre_attach(self: DataTree, parent: DataTree) -> None:
-        """
-        Method which superclass calls before setting parent, here used to prevent having two
-        children with duplicate names (or a data variable with the same name as a child).
-        """
-        super()._pre_attach(parent)
-        if self.name in list(parent.ds.variables):
-            raise KeyError(
-                f"parent {parent.name} already contains a data variable named {self.name}"
-            )
-
-    def to_dataset(self) -> Dataset:
+    def to_dataset(self, inherited: bool = True) -> Dataset:
         """
         Return the data in this node as a new xarray.Dataset object.
 
+        Parameters
+        ----------
+        inherited : bool, optional
+            If False, only include coordinates and indexes defined at the level
+            of this DataTree node, excluding inherited coordinates.
+
         See Also
         --------
         DataTree.ds
         """
         return Dataset._construct_direct(
-            self._variables,
-            self._coord_names,
-            self._dims,
+            self._variables if inherited else self._local_variables,
+            self._coord_names if inherited else self._local_coord_names,
+            self._dims if inherited else self._local_dims,
             self._attrs,
-            self._indexes,
+            self._indexes if inherited else self._local_indexes,
```

Comment:
The `.to_dataset` method becomes really simple, because the internal data model of the `DataTree` node is a close as possible to that of a `Dataset`.